File: learning-javascript/leet-code/src/jump-game/jump-game.ts
export function canJump(nums: number[]): boolean {
  let goal = nums.length - 1;
  for (let i = goal - 1; i >= 0; i--) {
    if (i + nums[i] >= goal) {
      goal = i;
    }
  }
  return goal === 0;
}

export function canJump2(nums: number[]): number {
	let jumps = 0;
	let l, r = 0;

	while (r < nums.length - 1) {
		let farthest = 0;
		for (let i = 0; i < r+1; i++) {
			farthest = Math.max(farthest, i + nums[i])
		}
		l = r + 1;
		r = farthest;
		jumps++ 
	}
	return jumps;
}

File: katas/jumpGame/jumpGame.py
def canJump(nums):
	goal = len(nums) - 1
	i = goal - 1
	while i >=0:	
		if i + nums[i] >= goal:
			goal = i
		i -= 1
	return goal == 0

def canJump2(nums):
	jumps = 0
	# boundary of our current level
	l = r = 0

	while r < len(nums) -1:
		farthest = 0
		for i in range(l, r+1):
			farthest = max(farthest, i + nums[i])
		l = r + 1
		r = farthest
		jumps +=1
	return jumps